miio/ihcooker.py

```diff
@@ -0,0 +1,746 @@
+import codecs
+import enum
+import logging
+import random
+import warnings
+from collections import defaultdict
+from typing import Dict, List, Optional, Union
+
+import click
+
+from .click_common import command, format_output
+from .device import Device
+from .exceptions import DeviceException
+
+_LOGGER = logging.getLogger(__name__)
+
+MODEL_EG1 = "chunmi.ihcooker.eg1"
+MODEL_EXP1 = "chunmi.ihcooker.exp1"
+MODEL_FW = "chunmi.ihcooker.chefnic"
+MODEL_HK1 = "chunmi.ihcooker.hk1"
+MODEL_KOREA1 = "chunmi.ihcooker.korea1"
+MODEL_TW1 = "chunmi.ihcooker.tw1"
+MODEL_V1 = "chunmi.ihcooker.v1"
+
+MODEL_VERSION1 = [MODEL_V1, MODEL_FW, MODEL_HK1, MODEL_TW1]
+MODEL_VERSION2 = [MODEL_EG1, MODEL_EXP1, MODEL_KOREA1]
+DEVICE_ID = {
+    MODEL_EG1: 2,  # TODO: TBD
+    MODEL_EXP1: 4,
+    MODEL_FW: 1,  # TODO: TBD
+    MODEL_HK1: 1,  # TODO: TBD
+    MODEL_KOREA1: 2,  # TODO: TBD
+    MODEL_TW1: 1,  # TODO: TBD
+    MODEL_V1: 1,
+}
+DEVICE_PREFIX = {k: "03%02d" % v for k, v in DEVICE_ID.items()}
+
+VERSION_1 = 1
+VERSION_2 = 2
+HARDWARE_VERSIONS = {
+    1: VERSION_1,
+    2: VERSION_2,  # Guess.
+    3: VERSION_2,  # Guess.
+    4: VERSION_2,
+    5: VERSION_2,  # Guess.
+}
+
+
+class IHCookerException(DeviceException):
+    pass
+
+
+class StageMode(enum.Enum):
+    TimerMode = 8
+    TemperatureMode = 2
+    FireMode = 0
+
+
+class OperationMode(enum.Enum):
+    Error = "error"
+    Finish = "finish"
+    Offline = "offline"
+    Pause = "pause"
+    Pause_time = "pause_time"
+    Precook = "precook"
+    Running = "running"
+    Set_01 = "set01"
+    Set_02 = "set02"
+    Set_03 = "set03"
+    Shutdown = "shutdown"
+    Timing = "timing"
+    Waiting = "waiting"
+
+
+def calcrc(data, n):
+    """
+     public static int calcrc(byte[] bArr, int i) {
+        int i2 = 0;
+        int i3 = 0;
+        while (true) {
+            i--;
+            if (i < 0) {
+                return 65535 & i2;
+            }
+            int i4 = i3 + 1;
+            int i5 = i2 ^ (bArr[i3] << 8);
+            for (int i6 = 0; i6 < 8; i6++) {
+                i5 = (32768 & i5) != 0 ? (i5 << 1) ^ 4129 : i5 << 1;
+            }
+            i2 = i5;
+            i3 = i4;
+        }
+    }
```

Comment:
This is unnecessary comment. Do you happen to know what sort of checksum function this is, if it's a common one this should simply be replaced with an invocation to an existing implementation to make the code more readable.

Reply:
This turned out to be CRC-CCITT, thanks to @nekromant  https://github.com/syssi/xiaomi_cooker/issues/15#issuecomment-848635066 .